pkg/controller/periodic/periodic.go

```diff
@@ -98,15 +106,46 @@ func (c *Controller) Gather() {
 		Steps:    threshold,
 		Cap:      interval,
 	}
-	for name := range c.gatherers {
+
+	// flush when all necessary gatherers were processed
+	defer func() {
+		if err := c.recorder.Flush(); err != nil {
+			klog.Errorf("Unable to flush the recorder: %v", err)
+		}
+	}()
+
+	var gatherersToProcess []gather.Interface
+
+	for _, gatherer := range c.gatherers {
+		if g, ok := gatherer.(gather.CustomPeriodGatherer); ok {
+			if g.ShouldBeProcessedNow() {
+				gatherersToProcess = append(gatherersToProcess, g)
+				g.UpdateLastProcessingTime()
+			}
+		} else {
+			gatherersToProcess = append(gatherersToProcess, gatherer)
+		}
+	}
+
+	var allFunctionReports []gather.GathererFunctionReport
+
+	for _, gatherer := range gatherersToProcess {
 		_ = wait.ExponentialBackoff(backoff, func() (bool, error) {
+			name := gatherer.GetName()
 			start := time.Now()
-			err := c.runGatherer(name)
+
+			ctx, cancel := context.WithTimeout(context.Background(), c.configurator.Config().Interval/2)
```

Comment:
Lol... so I was experimenting with setting some short timeout here (e.g 10s) and the log kinda exploded....especially due to two following gatherers:

- olm_operators
-  service_accounts

Reply:
I was experimenting with this again and maybe I would remove the extra logging statement in https://github.com/openshift/insights-operator/pull/399/files#diff-bdd45c95687562d23e5fdd43dea644d8fb86c95354192cedd72835743ba9fc0fR104. WDYT? I think the error should be obvious from a particular gatherer function and we shouldn't log it again....

I noticed it's almost the same now in https://github.com/openshift/insights-operator/blob/master/pkg/gather/clusterconfig/0_gatherer.go#L216, but the verbosity level is 5.

Reply:
But where else do we log it?

Reply:
The particular gatherer should log its own error IMO and it should also be visible in the metadata.

Reply:
But if we forget to log the error in the gatherer we're only gonna get it in the metadata

Reply:
Yes, what I am trying to say is that it should be gatherer responsibility (i.e. we shouldn't forget). The metadata is most important for me, because we usually don't have the IO log in our archive.

Reply:
Well, from my point of view it's better to do it once in the calling code than many times in gatherers. But I agree that metadata is more important than logs for us.